ast/src/parsed/display.rs

```diff
@@ -564,6 +552,121 @@ fn format_list<L: IntoIterator<Item = I>, I: Display>(list: L) -> String {
+    }
+
+    pub fn require_parentheses(&self) -> bool {
+        match self {
+            Expression::BinaryOperation(_, op, _) => {
+                op.associativity() == BinaryOperatorAssociativity::RequireParentheses
+            }
+            _ => false,
+        }
+    }
+
+    pub fn format_unary_operation(
+        &self,
+        op: &UnaryOperator,
+        exp: &Box<Expression<E>>,
+        f: &mut Formatter<'_>,
+    ) -> Result {
+        let _exp_string = if let (Some(_precision), Some(_exp_precision)) =
+            (self.precedence(), exp.precedence())
+        {
+            if _precision < _exp_precision {
+                format!("({})", exp)
+            } else {
+                format!("{}", exp)
+            }
+        } else {
+            format!("{}", exp)
+        };
+
+        if op.is_prefix() {
+            write!(f, "{}{}", op, _exp_string)
+        } else {
+            write!(f, "{}{}", _exp_string, op)
+        }
+    }
+
+    pub fn format_binary_operation(
+        left: &Box<Expression<E>>,
+        op: &BinaryOperator,
+        right: &Box<Expression<E>>,
+        f: &mut Formatter<'_>,
+    ) -> Result {
+        let use_left_parentheses = left.require_parentheses()
+            || match left.precedence() {
+                Some(left_precedence) => {
+                    left_precedence > op.precedence()
+                        || (left_precedence == op.precedence()
+                            && op.associativity() == BinaryOperatorAssociativity::Right)
+                }
+                None => false,
+            };
+
+        let use_right_parentheses = right.require_parentheses()
+            || match right.precedence() {
+                Some(right_precedence) => {
+                    right_precedence > op.precedence()
+                        || (right_precedence == op.precedence()
+                            && op.associativity() == BinaryOperatorAssociativity::Left)
+                }
+                None => false,
+            };
+
+        let left_string = if use_left_parentheses {
+            format!("({})", left)
+        } else {
+            format!("{}", left)
+        };
+        let right_string = if use_right_parentheses {
+            format!("({})", right)
+        } else {
+            format!("{}", right)
+        };
+
+        write!(f, "{} {} {}", left_string, op, right_string)
+    }
+    pub fn format_func_call(
+        fun_call: &FunctionCall<Expression<E>>,
+        f: &mut Formatter<'_>,
+    ) -> Result {
+        if let None = fun_call.function.precedence() {
+            write!(
+                f,
+                "{}({})",
+                fun_call.function,
+                format_list(&fun_call.arguments)
+            )
+        } else {
+            write!(
+                f,
+                "({})({})",
+                fun_call.function,
+                format_list(&fun_call.arguments)
+            )
+        }
+    }
+    pub fn format_index_access(
```

Comment:
Looks like this can just be the `impl<E: Display> Display for IndexAccess<E> {`, can't it?